VideoLocker/src/main/java/org/edx/mobile/third_party/iconify/IconDrawable.java

```diff
@@ -160,12 +177,25 @@ public int getIntrinsicWidth() {
 
     @Override
     public void draw(Canvas canvas) {
-        paint.setTextSize(getBounds().height());
+        Rect bounds = getBounds();
+        int height = bounds.height();
+        paint.setTextSize(height);
         Rect textBounds = new Rect();
         String textValue = valueOf(icon.character);
         paint.getTextBounds(textValue, 0, 1, textBounds);
-        float textBottom = (getBounds().height() - textBounds.height()) / 2f + textBounds.height() - textBounds.bottom;
-        canvas.drawText(textValue, getBounds().width() / 2f, textBottom, paint);
+        int textHeight = textBounds.height();
+        float textBottom = bounds.top + (height - textHeight) / 2f + textHeight - textBounds.bottom;
+        final boolean needMirroring = needMirroring();
+        if (needMirroring) {
+            canvas.save();
+            // Mirror the icon
+            canvas.translate(bounds.width(), 0);
+            canvas.scale(-1.0f, 1.0f);
+        }
+        canvas.drawText(textValue, bounds.exactCenterX(), textBottom, paint);
+        if (needMirroring) {
+            canvas.restore();
+        }
```

Comment:
@hanningni: `TextView` has lots of complex logic on it's `onDraw()` method as well :wink:. And if you check out the [`draw()` method][] of `BitmapDrawable`, you'll see that it uses pretty much the same logic. The canvas manipulations are necessary for mirroring, and in our case we need to check layout direction at that point as well since the `setLayoutDirection()` method is hidden (though you'll note that `BitmapDrawable` does the same thing also). I don't think a few get methods without any side effects will slow down the drawing much, and as for canvas changes, that's what the draw method of for :wink: